---
title: "SARS-CoV-2 Variants Coinfection"
date: "`r format(Sys.Date(), '%B %d, %Y')`"
output: 
  prettydoc::html_pretty:
    theme: leonids
    toc: yes
---
## Lineage assignment
### Pangolin and Nextclade 
 Lineage assigment using Pangolin and Nextclade of all samples
 ```{r, echo=FALSE}
 kable(params$lineage_assignment, "simple")
 ```
### Freyja
 
 Freyja plot shows the variant frequency for each sample. Samples suspected for coinfection of Delta and Omicron variants will show mixed variant prevalence.

 ![Freyja plot showing relative SARS-CoV-2 lineage abundance in the samples](`r params$freyja_plot`)

### VirStrain
 
 The samples were run against a database containing SARS-CoV-2 representative sequences for each variant. The VirStrain results shows the most possible lineage assignment for each sample. The figure and table below only shows samples that are suspected for coinfection.

 include_graphics(params$virstrain_html)

## Bammix flags for nucleotide mixture

 Bammix is a tool for plotting the nucleotide present 

 ![Bammix plot for nucleotide mixtures of flagged sample](`r params$bammix_plot`)

## Alternative allele fraction

### For each lineage-defining mutations
 
 ![Alternative allele fraction per mutation of sample flagged for SARS-CoV-2 coinfection](`r params$aafplot_mut`)

### For each amplicon
 ![Alternative allele fraction per amplicon](`r params$aafplot_amp`)
  
## References {.tabset .tabset-fade}

### Pipeline
The pipeline is available at [CoPi](https://github.com/lanadelrea/CoPi) by [Adeliza Realingo](https://github.com/lanadelrea). The tools used in this pipeline are as follows:

- [Pangolin](https://github.com/cov-lineages/pangolin) 

- [Nextclade](https://github.com/nextstrain/nextclade)

- [Freyja](https://github.com/andersen-lab/Freyja)

- [VirStrain](https://github.com/liaoherui/VirStrain)

- [Bammix](https://github.com/chrisruis/bammix)

- Custom scripts adopted from [Bolze et al.](doi: 10.1016/j.medj.2022.10.002) 

### Session Info
```{r, echo=FALSE}
sessionInfo()
```


